keycloak/keycloak-operator
# Support Pod-Resources in crd and statefulset

## JIRA ID
KEYCLOAK-12681

## Additional Information

I want to specify these requests and limits for all my pods in the cluster.

The crd ``` keycloaks.keycloak.org ``` is extended to support that for keycloak and for keycloak-postresql.

If requests are not set in the custom-resource, the operator provides a sane default for keycloak-memory and for keycloak-postgres memory and cpu.

## Verification Steps

Create a new keycloak cr without resources set:
```
kubectl apply -f -
apiVersion: keycloak.org/v1alpha1
kind: Keycloak
metadata:
  name: keycloak
  labels:
    app: sso
    test: sso
spec:
  instances: 1
  extensions:
    - https://github.com/aerogear/keycloak-metrics-spi/releases/download/1.0.4/keycloak-metrics-spi-1.0.4.jar
  externalAccess:
    enabled: False
  podDisruptionBudget:
    enabled: True
```
Check that keycloak and keycloak-postgres have been created without resources
```
kubectl get statefulsets.apps keycloak  -o jsonpath="{ ..resources }"  
map[] map[]✔ 

kubectl get deployment keycloak-postgresql  -o jsonpath="{ ..resources }"  
map[]✔ 
```



Create/Update the keycloak-cr with resources.
```
kubectl apply -f -
apiVersion: keycloak.org/v1alpha1
kind: Keycloak
metadata:
  name: keycloak
  labels:
    app: sso
    test: sso
spec:
  instances: 1
  extensions:
    - https://github.com/aerogear/keycloak-metrics-spi/releases/download/1.0.4/keycloak-metrics-spi-1.0.4.jar
  externalAccess:
    enabled: False
  podDisruptionBudget:
    enabled: True
  keycloakDeploymentSpec:
    resources:
      requests:
        cpu: 2500m
        memory: 700Mi
  postgresDeploymentSpec:
    resources:
      requests:
        cpu: 100m
        memory: 150Mi

```

Check the results (give it some time)
```
kubectl get statefulsets.apps keycloak  -o jsonpath="{ ..resources }"  
map[] map[requests:map[cpu:2500m memory:700Mi]]✔

kubectl get deployment keycloak-postgresql  -o jsonpath="{ ..resources }"  
map[requests:map[cpu:100m memory:150Mi]]✔```
```


Create/Update the keycloak-cr with resources and limits.

```
kubectl apply -f -
apiVersion: keycloak.org/v1alpha1
kind: Keycloak
metadata:
  name: keycloak
  labels:
    app: sso
    test: sso
spec:
  instances: 1
  extensions:
    - https://github.com/aerogear/keycloak-metrics-spi/releases/download/1.0.4/keycloak-metrics-spi-1.0.4.jar
  externalAccess:
    enabled: False
  podDisruptionBudget:
    enabled: True
  keycloakDeploymentSpec:
    resources:
      requests:
        cpu: 2700m
        memory: 800Mi
      limits:
        cpu: 2700m
        memory: 800Mi
  postgresDeploymentSpec:
    resources:
      requests:
        cpu: 250m
        memory: 250Mi
      limits:
        cpu: 250m
        memory: 250Mi
```

Check that the change has happened (give it some time)
```
kubectl get statefulsets.apps keycloak  -o jsonpath="{ ..resources }"  
map[] map[limits:map[cpu:2700m memory:800Mi] requests:map[cpu:2700m memory:800Mi]]

kubectl get deployment keycloak-postgresql  -o jsonpath="{ ..resources }"  
map[limits:map[cpu:250m memory:250Mi] requests:map[cpu:250m memory:250Mi]]✔ 
```

## Checklist:

## Additional Notes
